internal/collector/sysinfo/hardware/wayland_displays_linux.c

```diff
@@ -0,0 +1,170 @@
+#include "wayland_displays_linux.h"
+#include "wayland_displays_linux_test.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <wayland-client.h>
+
+static bool memory_error = false;
+
+static struct wayland_display **displays = NULL;
+static struct wl_output **outputs = NULL;
+
+static size_t count = 0;
+static size_t capacity = 4;
+
+static struct wl_display *display;
+static struct wl_registry *registry;
+
+// Function to handle geometry events from the Wayland output
+void handle_geometry(void *data, struct wl_output *output, int32_t x, int32_t y,
+                     int32_t physical_width, int32_t physical_height,
+                     int32_t subpixel, const char *make, const char *model,
+                     int32_t transform) {
+  struct wayland_display *info = (struct wayland_display *)data;
+  info->phys_width = physical_width;
+  info->phys_height = physical_height;
+}
+
+// Function to handle mode events from the Wayland output
+void handle_mode(void *data, struct wl_output *output, uint32_t flags,
+                 int32_t width, int32_t height, int32_t refresh) {
+  struct data_info *data_i = data;
+
+  if (flags & WL_OUTPUT_MODE_CURRENT) {
+    struct wayland_display *info = (struct wayland_display *)data;
+    info->width = width;
+    info->height = height;
+    info->refresh = refresh;
+  }
+}
+
+static const struct wl_output_listener output_listener = {
+    .geometry = handle_geometry,
+    .mode = handle_mode,
+    .done = NULL,
+    .scale = NULL,
+};
+
+// Function to handle global events from the Wayland registry
+void global_handler(void *data, struct wl_registry *registry, uint32_t name,
+                    const char *interface, uint32_t version) {
+  if (strcmp(interface, "wl_output") != 0) {
+    return;
+  }
+  // Check if we need to increase the capacity of the arrays
+  if (count >= capacity) {
+    capacity *= 2;
+    struct wl_output **new_outputs =
+        realloc(outputs, capacity * sizeof(struct wl_output *));
+
+    struct wayland_display **new_displays =
+        realloc(displays, capacity * sizeof(struct wayland_display *));
+    if (!new_outputs || !new_displays) {
+      free(new_outputs);
+      free(new_displays);
+      new_displays = NULL;
+      new_outputs = NULL;
+      memory_error = true;
+      return;
+    }
+    outputs = new_outputs;
+    displays = new_displays;
+  }
+
+  struct wayland_display *display = malloc(sizeof(struct wayland_display));
+  if (!display) {
+    memory_error = true;
+    return;
+  }
+
+  struct wl_output *output =
+      wl_registry_bind(registry, name, &wl_output_interface, 1);
+
+  displays[count] = display;
+  outputs[count] = output;
+  count++;
+
+  wl_output_add_listener(output, &output_listener, display);
+}
+
+void global_remove(void *data, struct wl_registry *registry, uint32_t name) {
+  free(data);
```

Comment:
Calling `free(data);` here is essentially calling `free(NULL);` because this data is the value passed to `wl_registry_add_listener` at line 120:

```
wl_registry_add_listener(registry, &registry_listener, NULL);
```

Instead we can make this function empty ;)